pkg/api/store/cluster/cluster_store.go

```diff
@@ -17,3 +25,66 @@ func (r *Store) ByID(apiContext *types.APIContext, schema *types.Schema, id stri
 	}
 	return r.Store.ByID(apiContext, schema, id)
 }
+
+func (r *Store) Create(apiContext *types.APIContext, schema *types.Schema, data map[string]interface{}) (map[string]interface{}, error) {
+
+	name, ok := data["name"].(string)
+	if !ok {
+		return nil, httperror.NewFieldAPIError(httperror.MissingRequired, "Cluster name", "")
+	}
+
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	if err := canUseClusterName(apiContext, name); err != nil {
+		return nil, err
+	}
+
+	return r.Store.Create(apiContext, schema, data)
+}
+
+func (r *Store) Update(apiContext *types.APIContext, schema *types.Schema, data map[string]interface{}, id string) (map[string]interface{}, error) {
+
+	existingCluster, err := r.ByID(apiContext, schema, id)
+
+	if err != nil {
+		return nil, err
+	}
+
+	clusterName, ok := existingCluster["name"].(string)
+	if !ok {
+		clusterName = ""
+	}
+
+	updatedName, ok := data["name"].(string)
```

Comment:
do we always expect the name in the update request? does UI send it all the time? if so, may be it's something we can check at the very top of the method, to save on querying the existing cluster. Also it can be extracted to its own method, so the call can replace line 31-34 in create

Reply:
Yeah name is always sent in the data[] for the UPdate call, infact all properties are present in the data[]. So to know if name is getting updated, we would have to compare with the existing name.

Reply:
Will swap the order of checking cluster name and pulling existing object. But just checking if clusterName is missing need not be a common method.